### PR TITLE
Add Vision API Face detection.

### PR DESCRIPTION
**Pull request by daspecster**
Initial add for face detection.

I'm not sure about the way I have `Likelihood` and `FaceLandmarkTypes` setup.
If you have some ideas on how that could be improved I would love to hear them.

This looks large, but it's mostly DTOs for the API response.

LMKWYT!

### CODE REVIEW

**Comment by tseaver** [comment]
@daspecster pylint is [unhappy](https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/156342391#L478-L495).

**Comment by daspecster** [comment]
Hmm I ran lint locally and it passed before I pushed...that's annoying.

**Comment by daspecster** [comment]
@tseaver how should I rename those attributes?
I feel like those are pretty good. Right now I'm thinking `x_coordinate` but that makes accessing them kind of weird. `face.landmarks.left_eye.x_coordinate`.

**Comment by dhermes** [comment]
@daspecster Did you run `tox -e lint` or something else? The names are fine. Just use a 

```python
# pylint: disable=invalid-name
# code that violates goes here
# pylint: enable=invalid-name
```

**Comment by daspecster** [comment]
@dhermes I can still do that if we want. I looked around and renaming to `x_coordinate` seemed to be how most people have solved this issue.

**Comment by tseaver** [comment]
@daspecster Needs a rebase after #2227,

**Comment by daspecster** [comment]
@tseaver sorry about that, I can't keep track of my spinning plates.

There was a PR before that made a conflict that I had merged in so I thought that was the one you were talking about.

File: docs/vision-usage.rst [hunk 1]
```diff
@@ -38,31 +38,30 @@ Annotate a single image
 
 .. code-block:: python
 
+    >>> import io
     >>> from google.cloud import vision
     >>> client = vision.Client()
-    >>> image = client.image('./image.png')
+    >>> with io.open('./image.png', 'rb') as image_file:
+    >>>     image = client.image(image_file.read())
```
Comment (tseaver) [inline docs/vision-usage.rst]: Use `...` for lines within a block.

File: docs/vision-usage.rst [hunk 2]
```diff
@@ -38,31 +38,30 @@ Annotate a single image
...
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
 .. code-block:: python
 
-     >>> first_image = client.image('./image.jpg')
+     >>> import io
+     >>> from gcloud import vision
+     >>> client = vision.Client()
+     >>> with io.open('./image.png', 'rb') as image_file:
+     >>>     first_image = client.image(image_file.read())
```
Comment (tseaver) [inline docs/vision-usage.rst]: Ditto.

File: docs/vision-usage.rst [hunk 3]
```diff
@@ -86,9 +85,13 @@ You can call the detection method manually.
 .. code-block:: python
 
     >>> from google.cloud import vision
+    >>> from google.cloud.vision.image import Feature
+    >>> from google.cloud.vision.image import FeatureTypes
     >>> client = vision.Client()
     >>> image = client.image('gs://my-test-bucket/image.jpg')
-    >>> faces = image.detect(type=vision.FACE_DETECTION, limit=10)
+    >>> features = [Feature(FeatureTypes.FACE_DETECTION, 5),
+                   Feature(FeatureTypes.LOGO_DETECTION, 3)]
```
Comment (tseaver) [inline docs/vision-usage.rst]: Use `...`.

File: google/cloud/vision/client.py [hunk 4]
```diff
@@ -98,12 +98,22 @@ def annotate(self, image, features):
...
+    def image(self, image_source):
+        """Get instance of Image using current client.
+
+        :type image_source: str or bytes
+        :param image_source: Byte stream of an image or a GCS URI.
+
+        :rtype: :class:`google.cloud.vision.image.Image`
+        :returns: Image instance with the current client attached.
+        """
+        return Image(client=self, image_source=image_source)
```
Comment (tseaver) [inline google/cloud/vision/client.py]: I'm still finding the overload of `bytes` and `str` unpleasant:  I think there should be separate parameters (or maybe even separate factories) for the two cases (one with raw image bytes in hand and one with a GCS URI).

Comment (daspecster) [inline google/cloud/vision/client.py]: Should the public method change too?
`Client.image_from_bytes()`, `Client.image_from_gcs()`?

Or in `Client.image()` I can just pick which Image factory to return, but I would have to add the logic to determine if the string is a GCS URI here in `Client.image()`.

Comment (tseaver) [inline google/cloud/vision/client.py]: I would be fine with just having separate parameters for the two usecases:

```python
image1 = client.image(content=BYTES)
image2 = client.image(source_uri='gs://my-bucket/image.jpg')
```

And likewise for `Image.__init__`.

**Comment by daspecster** [comment]
I'm not sure what's going on but for some reason `tox -e cover` isn't finding the py file for `test_image.py`.

```bash
unit_tests/vision/test_client.py .......
unit_tests/vision/test_connection.py .
unit_tests/vision/test_face.py ......
unit_tests/vision/test_feature.py ...
unit_tests/vision/test_image.py ...Coverage.py warning: Module google.cloud has no Python source.
```

**Comment by dhermes** [comment]
@daspecster The line

```
unit_tests/vision/test_image.py ...Coverage.py warning: Module google.cloud has no Python source
```

has two different messages, separate from one another (essentially there is a missing newline, or more likely one went to stdout and one to stderr).

The statement `Module google.cloud has no Python source` just reflects that `google.cloud.__init__` never gets imported since it's a namespace package.

**Comment by daspecster** [comment]
Hmmm, this is the final output I get.

```bash
unit_tests/vision/test_feature.py                                                                         23      0      0      0   100%
unit_tests/vision/test_image.py                                                                           34      0      0      0   100%
--------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                  34899      0   2687      1    99%

FAIL Required test coverage of 100% not reached. Total coverage: 100.00%
```

**Comment by daspecster** [comment]
Ah I found it! Thank you!

**Comment by dhermes** [comment]
Yeah the 99% comes from another module under 100%

**Comment by dhermes** [comment]
```
.tox/cover/lib/python2.7/site-packages/google/cloud/vision/image.py                                       31      0      8      1    97%   46->
```

**Comment by daspecster** [comment]
[This](https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2224/files#diff-87d25e0bd2187482b5fd4412dbfe435eR34) should be covering line 46 of image.py.

**Comment by dhermes** [comment]
It's a branch miss.

**Comment by daspecster** [comment]
A single test shouldn't have to test all branches in a method though? In this case one test covered one branch and another test covered the other AFAICT.

**Comment by tseaver** [comment]
The `content` / `source_uri` changes SGTM.  Merge when Travis is green.

**Comment by dhermes** [comment]
> A single test shouldn't have to test all branches in a method though? In this case one test covered one branch and another test covered the other AFAICT.

You can use as many test cases as you need to test all lines and all branches introduced.

**Pull request merged**